2017/proposals/bansal-utkarsh.md

```diff
@@ -0,0 +1,628 @@
+code coverage running as usual. This is being done in a single PR to avoid any weird coexistence of two testing 
+libraries.
+
+* Modify the code to make it discoverable by pytest (take care of `__init__` and inheritance problems as explained above)
+* Port the 5 existing nose plugins to pytest. Pytest has awesome documentation on how to implement plugins which can be
+ found [here](http://doc.pytest.org/en/latest/writing_plugins.html) .
+    * Capture error (`capture_err.py`) This plugin captures stderr during test execution. If the test fails
+    or raises an error, the captured output is appended to the error or failure output. Pytest already has this
+feature inbuilt so this plugin can be dropped.
+
+        **Example**
+        ```python
+        def check_even(n, nn):
+            print(n)
+            assert n % 2 == 0 or nn % 2 == 0
+
+
+        def test_evens():
+            for i in range(0, 5):
+                yield check_even, i, i * 3
+        ```
+
+        **Output**
+        ```
+        ======================================= FAILURES =======================================
+        ____________________________________ test_evens[1] _____________________________________
+
+        n = 1, nn = 3
+
+            def check_even(n, nn):
+                print(n)
+        >       assert n % 2 == 0 or nn % 2 == 0
+        E       assert ((1 % 2) == 0 or (3 % 2) == 0)
+
+        experiments.py:37: AssertionError
+        --------------------------------- Captured stdout call ---------------------------------
+        1
+        ____________________________________ test_evens[3] _____________________________________
+
+        n = 3, nn = 9
+
+            def check_even(n, nn):
+                print(n)
+        >       assert n % 2 == 0 or nn % 2 == 0
+        E       assert ((3 % 2) == 0 or (9 % 2) == 0)
+
+        experiments.py:37: AssertionError
+        --------------------------------- Captured stdout call ---------------------------------
+        3
+        ```
+
+
+    * Cleanup (`cleanup.py`) This plugin deletes XDR offset files after all testing is done. This will have to
+    be re-written as a pytest plugin.
+    * Knownfailure (`knownfailure.py`) This is a decorator to check tests that are expected to fail. Pytest has
+    something similar to this called `xfail`. The documentation for this is available [here](http://doc.pytest.org/en/latest/skipping.html#mark-a-test-function-as-expected-to-fail) .
+    This needs discussion to finalize if we need to port the existing code or if `xfail` is good enough for our use.
+    * Memory leak (`memleak.py`) This plugin detects memory leaks in tests. There is an open-source (MIT license) plugin, [pytest-leaks](https://github.com/abalkin/pytest-leaks).
+    This also needs discussion to finalize if we need to port the existing code or if pytest-leaks is good enough for our use.
```

Comment:
Yes, thanks for that! I had a look. I think I'll write down all the three alternatives and this can further be discussed/profiled. Would that be okay?